Documentation for .NET Core Uninstall Tool

sfoslund opened the pull request:
## Summary

This is replacing a previous PR as the owner is no longer at Microsoft. See #13562.
Added documentation for .NET Core Uninstall Tool (dotnet-core-uninstall). Following the suggestions in #13529, the documentation is now standalone with a link from How to remove the .NET Core Runtime and SDK.

yuchong-pan committed:
Added documentation for .NET Core Uninstall Tool

yuchong-pan committed:
Revised .NET Core Uninstall Tool documentation for wordings

yuchong-pan committed:
Removed locale parameters from links

yuchong-pan committed:
Fixed a typo in .NET Core Uninstall Tool documentation

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/index.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Fixed numberings in the documentation

yuchong-pan committed:
Removed --do-it option and added --dry-run and --yes/-y options

yuchong-pan committed:
Updated examples with new options and re-ordering

yuchong-pan committed:
Added section for VS dependencies on .NET Core SDKs

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Corrected information on macOS

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Fixed typos

yuchong-pan committed:
Removed locale code from URLs

yuchong-pan committed:
Added explanation of NuGetFallbackFolder

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/additional-tools/index.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/versions/remove-runtime-sdk-versions.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Update docs/core/versions/remove-runtime-sdk-versions.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Fixed syntax errors

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Fixed syntax errors

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Fixed syntax errors

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Added paragraphs for NuGetFallbackFolder

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Added small changes

yuchong-pan committed:
Merge branch 'dotnet-core-uninstall-docs' of https://github.com/yuchong-pan/docs into dotnet-core-uninstall-docs

yuchong-pan committed:
Added small changes

yuchong-pan committed:
Fixed some issues with options

yuchong-pan committed:
Changed the wording from 'uninstall' to 'remove'

yuchong-pan committed:
Update docs/core/additional-tools/dotnet-core-uninstall.md

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

yuchong-pan committed:
Changed the wording from 'remove' to 'delete' for the tool itself

yuchong-pan committed:
Added an example with response files

sfoslund committed:
Updating docs for command restructure of uninstall tool, updating based on PR feedback

sfoslund commented:
These changes have been discussed with @KathleenDollard and @wli3.

KathleenDollard commented on docs/core/additional-tools/dotnet-core-uninstall.md:
```diff
@@ -0,0 +1,332 @@
+## How to use the tool
+
+1. Download and install the tool.
+2. Display installed .NET Core SDKs and Runtimes that can be removed by this tool.
```
```suggestion
2. Display installed .NET Core SDKs and Runtimes that can be removed by this tool using `dotnet-core-uninstall list`.
```

KathleenDollard commented on docs/core/additional-tools/dotnet-core-uninstall.md:
```diff
@@ -0,0 +1,332 @@
+
+1. Download and install the tool.
+2. Display installed .NET Core SDKs and Runtimes that can be removed by this tool.
+3. Do a dry run to ensure the right things will be uninstalled, and uninstall .NET Core SDKs and Runtimes by running the tool with Admin privileges.
```
```suggestion
3. Use `dotnet-core-uninstall dry-run` to ensure the right things will be uninstalled.
4. Use `dotnet-core-uninstall remove` to uninstall .NET Core SDKs and Runtimes. This must be run with Admin privileges.
```

KathleenDollard commented on docs/core/additional-tools/dotnet-core-uninstall.md:
```diff
@@ -0,0 +1,332 @@
+
+### Uninstalling .NET Core SDKs and Runtimes
+
+`dotnet-core-uninstall dry-run` and `dotnet-core-uninstall whatif` display the .NET Core SDKs and Runtimes that will be removed based on the options provided without performing the uninstall. These commands are equivalent.
```
```suggestion
`dotnet-core-uninstall dry-run` and `dotnet-core-uninstall whatif` display the .NET Core SDKs and Runtimes that will be removed based on the options provided without performing the uninstall. These commands are synonyms.
```

KathleenDollard commented on docs/core/additional-tools/dotnet-core-uninstall.md:
```diff
@@ -0,0 +1,332 @@
+
+`dotnet-core-uninstall dry-run` and `dotnet-core-uninstall whatif` display the .NET Core SDKs and Runtimes that will be removed based on the options provided without performing the uninstall. These commands are equivalent.
+
+Since this is a destructive tool, it is **highly** recommended that you run one of these commands before the remove command to determine the .NET Core SDKs and Runtimes that will be removed. [This article explains which SDKs and Runtimes are safe to remove.](/versions/remove-runtime-sdk-versions#should-i-remove-a-version)
```
```suggestion
Since this is a destructive tool, it is **highly** recommended that you do a dry run before the running the remove command. This will tell you what .NET Core SDKs and Runtimes that will be removed. [This article explains which SDKs and Runtimes are safe to remove.](/versions/remove-runtime-sdk-versions#should-i-remove-a-version)
```

KathleenDollard reviewed (commented):
@mairaw is combining the switches for dry-run and remove OK, or do we need to repeat them? (we restructured the command since you last looked at this.

sfoslund committed:
Apply Kathleen's suggestions from code review

Co-Authored-By: Kathleen Dollard <kathleen.dollard@microsoft.com>

sfoslund committed:
Fix Markdown issues from lint

sfoslund committed:
Updating release page, uninstall directions, and examples

sfoslund commented:
I updated these changes so we're pointing to the correct download page and the directions for uninstalling the tool are up to date. 
@wli3 the directions for uninstalling the tool are [here](https://github.com/dotnet/docs/pull/15288/files#diff-96c15e624d80eede2f6cb0e01840490aR346), in case you want to add anything. 
@KathleenDollard I added notes [here](https://github.com/dotnet/docs/pull/15288/files#diff-96c15e624d80eede2f6cb0e01840490aR290) and [here](https://github.com/dotnet/docs/pull/15288/files#diff-96c15e624d80eede2f6cb0e01840490aR130) to make it clear that "all" doesn't necessarily mean all. I'd like your feedback on the wording if possible.

sfoslund committed:
Adding tab ending to uninstall directions

sfoslund committed:
Updating heading levels for markdown linting

mairaw committed:
edits

mairaw commented on docs/core/additional-tools/dotnet-core-uninstall.md:
```diff
@@ -0,0 +1,357 @@
+
+The tool supports Windows and macOS. Linux is currently not supported.
+
+Note that the Windows version of this tool can only uninstall .NET Core SDKs and runtimes that were installed using .NET Core SDK and runtime installers. Prior to Visual Studio 2019 Update 16.3, the Visual Studio installer called the .NET Core SDK installer and these versions can be uninstalled with the tool. The `dotnet-core-uninstall list` command displays which SDKs can be uninstalled with the ...
```
What happens starting with 16.3?

sfoslund replied:
@wli3 would know more than me

mairaw replied:
@wli3 can you please comment?

wli3 replied:
After that Visual Studio is in charge of its own copy. That means Visual Studio will bring a copy of dotnet core, and the user cannot uninstall it from "add remove programs". And Visual Studio will also update that copy in VS update. It will also go away when VS is uninstalled

mairaw committed:
update metadata

mairaw commented:
I've made some edits to the first article and I have more to make. I'll also leave some comments for us to discuss the approach in some areas.

mairaw commented:
For author, it would be good if it is someone currently on the team since the person gets pinged whenever a new issue is opened. We can add a byline to this article to indicate who worked on this.

mairaw commented on docs/core/additional-tools/dotnet-core-uninstall.md:
```diff
@@ -0,0 +1,357 @@
+
+## How to use the tool
+
+1. Display installed .NET Core SDKs and runtimes that can be removed by this tool using `dotnet-core-uninstall list`.
```
it felt weird to have install and uninstall steps in the how to use the tool guide.

Also, I'm wondering if each one of these items should be simplified and be a link to the section below or if we should simply skip and have the steps directly in the header such as in https://docs.microsoft.com/en-us/visualstudio/install/install-visual-studio?view=vs-2019

I'm also thinking that usage instructions and command syntax should be broken down into different sections.

sfoslund replied:
Having the steps directly in the header would make sense if we're splitting the options for `dotnet-core-uninstall dry-run` and `dotnet-core-uninstall remove` into two different sections. Is that what you mean by braking down the usage instructions or did you have something else in mind?

mairaw replied:
I think install, use, and uninstall should be separate instructions instead of having install and uninstall as steps of how to use the tool - that seems wrong. I tried to make that more clear in the latest updates - see what you think! But I think what I'm also struggling with on this particular article is that is a mix of reference (with the commands syntax) and how to. 
I see two options:
- We merge as is (if you're ok with my changes) and watch for customer's feedback.
- We separate the commands syntax and the prescriptive guidance into different sections (or even different articles). We could even merge this and then I open a new PR with a proposal on top of this.

sfoslund replied:
It's up to you, but the changes you made look good to me and I am okay with merging this if @KathleenDollard is.

mairaw commented on docs/core/additional-tools/dotnet-core-uninstall.md:
```diff
@@ -0,0 +1,357 @@
+
+Use the `dotnet-core-uninstall list` command lists installed .NET Core SDKs and runtimes that can be removed with this tool. Some SDKs and runtimes may be required by Visual Studio and they're displayed with a note of why it isn't recommended to uninstall them.
+
+On Windows, the tool can only uninstall SDKs and runtimes that were installed using the .NET Core SDK and runtime installer, or the Visual Studio installer prior to Visual Studio 2019 16.3. On macOS, the tool can only uninstall SDKs and runtimes located in the */usr/local/share/dotnet* folder. This tool doesn't work on Linux. Because of these limitations, the tool may not be able to uninstall all ...
```
this info about 16.3 and that it doesn't work on Linux was already mentioned in the intro. We should avoid repeating the same info so we should choose where it makes more sense to have this.

sfoslund replied:
Probably makes sense to have it in the intro then.

mairaw committed:
Rename dotnet-core-uninstall.md to uninstall-tool.md

sfoslund commented on docs/core/additional-tools/uninstall-tool.md:
```diff
@@ -0,0 +1,357 @@
+
+* **`--aspnet-runtime`**
+
+  Lists all the ASP.NET Core runtimes that are installed on your machine.
```
The tool does not necessarily list all runtimes/sdks installed on the machine- for example it doesn't list bundles installed via zips. More accurate wording here would be that it lists all runtimes that are able to be uninstalled by this tool.

mairaw replied:
Ok, makes sense.

sfoslund committed:
Updating uninstall tool docs based on PR feedback

mairaw committed:
more edits

mairaw committed:
fix broken link

mairaw committed:
update ms.date

mairaw committed:
remove bad character

mairaw committed:
edits

mairaw committed:
roll back ms.custom

mairaw committed:
fix anchor name

mairaw committed:
move entry up

mairaw committed:
Merge branch 'master' into dotnet-core-uninstall-docs

mairaw committed:
add TOC entry

mairaw committed:
test zones

mairaw committed:
rearrange headings

mairaw committed:
remove extra line

mairaw committed:
fix formstting

mairaw committed:
remove zones

sfoslund commented:
@mairaw I just chatted with @KathleenDollard in person and she is okay with merging this if you are. Do you have anything else you need from us before merging?

mairaw committed:
Update uninstall-tool.md

mairaw committed:
add tracking metadata

mairaw commented:
@sfoslund sounds good. I made some tiny changes now and I think we can merge. Once I'm back in January, I'd like to see if structuring this article a bit differently could yield a better format.

mairaw committed:
add mini script with links

mairaw committed:
add note about VS 16.3+

shanselman commented:
Is this tool gone? I followed a link to the .NET site and it was a 404. @mairaw

wli3 commented:
https://dotnet.microsoft.com/download/dotnet-core/uninstall-tool is not put up yet. It is still in prerelease phrase. @mairaw @sfoslund 

https://github.com/dotnet/cli-lab/releases is the only link today

shanselman commented:
Cool thanks @wli3! It works great
![image](https://user-images.githubusercontent.com/2892/74887149-f9a28480-532e-11ea-948d-06bdb594af9f.png)

mairaw commented:
@shanselman did you see the link on this PR or somewhere else? I did fix that broken link later on with PR #16743

shanselman commented:
Shoot, I'm not sure how I got there. Perhaps from an issue somewhere in
here. If it'll go up sometime soon it doesn't really matter.

On Wed, Feb 19, 2020 at 5:05 PM Maira Wenzel <notifications@github.com>
wrote:

> @shanselman <https://github.com/shanselman> did you see the link on this
> PR or somewhere else? I did fix that broken link later on with PR #16743
> <https://github.com/dotnet/docs/pull/16743>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dotnet/docs/pull/15288?email_source=notifications&email_token=AAAAWTAKJS4CCJSEYI2X653RDXJMTA5CNFSM4JCMGLWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMKKU5I#issuecomment-588556917>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAWTE5B2MDTUOSMSFTEIDRDXJMTANCNFSM4JCMGLWA>
> .
>

mairaw commented:
No worries. Just wanted to make sure we didn't have anything else broken on docs.
